src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs

```diff
@@ -350,4 +370,205 @@ internal static bool TryParseSeverity(string value, out ReportDiagnostic severit
             return false;
         }
     }
+
+    /// <summary>
+    /// Represents a global key that was unset due to conflicts
+    /// </summary>
+    public readonly struct AnalyzerUnsetKey
+    {
+        /// <summary>
+        /// The name of the key that was unset
+        /// </summary>
+        public string KeyName { get; }
+
+        /// <summary>
+        /// The section this key came from 
+        /// </summary>
+        /// <remarks>
+        /// Will be 'Global Section' when <see cref="IsGlobalSection"/> is <c>true</c>.
+        /// </remarks>
+        public string SectionName { get; }
+
+        /// <summary>
+        /// <c>true</c> if the section this key came from was the global section
+        /// </summary>
+        public bool IsGlobalSection { get; }
+
+        /// <summary>
+        /// The list of paths of the global configs that each set this key
+        /// </summary>
+        public ImmutableArray<string> ConfigPaths { get; }
+
+        internal AnalyzerUnsetKey(string keyName, string sectionName, bool isGlobalSection, ImmutableArray<string> paths)
+        {
+            KeyName = keyName;
+            SectionName = sectionName;
+            IsGlobalSection = isGlobalSection;
+            ConfigPaths = paths;
+        }
+    }
+
+    /// <summary>
+    /// Builds a global analyzer config from a series of configs
+    /// </summary>
+    internal class GlobalAnalyzerConfigBuilder
+    {
+        internal const string GlobalKey = "is_global";
+
+        PooledDictionary<string, ImmutableDictionary<string, (string, string)>.Builder> _values;
+        PooledDictionary<string, PooledDictionary<string, ArrayBuilder<string>>> _duplicates;
+
+        private GlobalAnalyzerConfigBuilder()
+        {
+            _values = PooledDictionary<string, ImmutableDictionary<string, (string, string)>.Builder>.GetInstance();
+            _duplicates = PooledDictionary<string, PooledDictionary<string, ArrayBuilder<string>>>.GetInstance();
+        }
+
+        /// <summary>
+        /// Filter out and merge any global configs
+        /// </summary>
+        /// <param name="analyzerConfigs">A list of <see cref="AnalyzerConfig"/> to filter out </param>
+        /// <param name="unsetAnalyzerKeys">Returns any global keys that were unset due to duplication</param>
+        /// <returns>An <see cref="ImmutableArray{AnalyzerConfig}"/> that contains the non-global configs and a single merged global config</returns>
+        internal static ImmutableArray<AnalyzerConfig> FilterGlobalConfigs<TList>(TList analyzerConfigs, out ImmutableArray<AnalyzerUnsetKey> unsetAnalyzerKeys) where TList : IReadOnlyCollection<AnalyzerConfig>
+        {
+            // don't do any actual work if there aren't any global configs
+            if (!containsGlobalConfigs())
+            {
+                unsetAnalyzerKeys = ImmutableArray<AnalyzerUnsetKey>.Empty;
+                return analyzerConfigs.ToImmutableArray();
```

Comment:
Given the call site and this particular behavior wouldn't this method be better authored if it just took an `ImmutableArray` as the parameter?

Reply:
I ended up re-factoring this whole method to be much cleaner.

---
In reply to: [421684651](https://github.com/dotnet/roslyn/pull/43889#discussion_r421684651) [](ancestors = 421684651)